keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,252 @@
+    Name: "some_deprecated_metric",
+    Help: "some description",
+    DeprecatedVersion: "1.15", // this is a custom metadata field
+}
+
+var alphaMetricDefinition = kubemetrics.CounterOpts{
+    Name: "some_alpha_metric",
+    Help: "some description",
+    StabilityLevel: kubemetrics.ALPHA, // this is also a custom metadata field
+}
+```
+
+### Metric Instantiation Phase
+
+Currently, the metric instantiation phase looks like this:
+
+```go
+var someCounterVecMetric = prometheus.NewCounterVec(
+    someMetricDefinition,
+    []string{"some-label", "other-label"},
+}
+```
+
+Wrapping the prometheus constructors would allow us to take, as inputs, the modified metric definitions defined above, returning a custom kubernetes metric object which contains the metric which would have been instantiated as well as the custom metadata:
+
+```go
+var deprecatedMetric = kubemetrics.NewCounterVec( // this is a wrapped initializer, which takes in our custom metric definitions
+    deprecatedMetricDefinition, // this is our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+var alphaMetric = kubemetrics.NewCounterVec{
+    alphaMetricDefinition, // this is also our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+```
+
+### Metric Enrollment Phase
+
+Currently, metric enrollment involves calls to a prometheus function which enrolls the metric in a global registry, like so:
+```go
+prometheus.MustRegister(someCounterVecMetric)
+```
+
+Wrapping a prometheus registry with a kubernetes specific one, would allow us to take our custom metrics from our instantiation phase and execute custom logic based on our custom metadata. Our custom registry would hold a reference to a prometheus registry and defer metric enrollment unless preconditions were met:
+
+```go
+import version "k8s.io/apimachinery/pkg/version"
+
+type Registry struct {
+    promregistry *prometheus.Registry
+    KubeVersion version.Info
+}
+
+// inject custom registration behavior into our registry wrapper
+func (r *Registry) MustRegister(metric kubemetrics.Metric) {
+    // pretend we have a version comparison utility library
+    if metricutils.compare(metric.DeprecatedVersion).isLessThan(r.KubeVersion) {
+        // check if binary has deprecated metrics enabled otherwise
+        // no-op registration
+        return
+    } else if metricutils.compare(metric.DeprecatedVersion).isEqual(r.KubeVersion) {
+        // append deprecated text to description
+        // emit warning in logs
+        // continue to actual registration
+    }
+    // append alpha text to metric description if metric.isAlpha
+    // fallback to original prometheus behavior
+    r.promregistry.MustRegister(metric.realMetric)
+}
+
+```
+
+Which we would invoke, like so:
+```go
+kubemetrics.MustRegister(deprecatedMetric)
+kubemetrics.MustRegister(alphaMetric)
+```
+
+## Stability Classes
```

Comment:
Similar to the `--enable-hard-deprecated-metrics` flag, will there be the option to disable all or specific alpha metrics?  I could easily see wanting to disable a metric because it introduced a performance regression, for example.

Reply:
Yeah, actually I had originally meant to include that.

Reply:
Oh, actually I did already include this. This is in the unresolved questions section. But since that was a source of confusion, I linked it to the alpha discussion and referenced it at the bottom as a footnote.